src/app/components/table-of-content/table-of-content.component.ts

```diff
@@ -18,13 +26,71 @@ import { NgClass } from '@angular/common';
   templateUrl: './table-of-content.component.html',
   styleUrl: './table-of-content.component.scss',
 })
-export class TableOfContentComponent {
+export class TableOfContentComponent implements AfterViewInit {
   @Input() set headers(headers: HeaderNode[]) {
     this.dataSource.data = headers;
     this.treeControl.expandAll();
+    if (headers.length) {
+      this.createOffsets(headers);
+      this.checkFirstHeader();
+    }
   }
+
   @Input() current!: string;
 
+  offsetHeaders: string[] = [];
+
+  activeHeader = '';
+
+  @HostListener('document:scroll', ['$event'])
+  public onViewportScroll() {
+    if (this.offsetHeaders.length) {
+      const scroll = Math.ceil(this.document.documentElement.scrollTop);
+      const fullHeightWithScroll = this.document.body.scrollHeight;
+      const fullHeight =
+        this.document.body.scrollHeight - this.document.body.clientHeight;
+
+      for (let i = 0; i < this.offsetHeaders.length; i++) {
+        const element = this.offsetHeaders[i];
+        const nextElement = this.offsetHeaders[i + 1];
+
+        const offset = this.document.getElementById(element)?.offsetTop;
+        const nextOffset = nextElement
+          ? this.document.getElementById(nextElement)?.offsetTop
+          : fullHeightWithScroll;
+
+        if (offset == null || nextOffset == null) {
+          continue;
+        }
+
+        if (fullHeight === scroll) {
+          this.activeHeader = this.offsetHeaders[this.offsetHeaders.length - 1];
+          return;
+        }
+
+        if (scroll >= offset && scroll < nextOffset) {
+          this.activeHeader = element;
+          return;
+        }
+      }
+    }
+  }
```

Comment:
We can simplify this method a bit using Array functions. This way we keep it immutable as well
```suggestion
  activeHeader: string | undefined = '';

  @HostListener('document:scroll', ['$event'])
  public onViewportScroll() {
    if (this.offsetHeaders.length) {
      const scroll = Math.ceil(this.document.documentElement.scrollTop);
      const fullHeightWithScroll = this.document.body.scrollHeight;
      const fullHeight =
        this.document.body.scrollHeight - this.document.body.clientHeight;

      this.activeHeader = this.offsetHeaders.find((element, i, offsetHeaders) => {
        const nextElement = offsetHeaders[i + 1];

        const offset = this.document.getElementById(element)?.offsetTop;
        const nextOffset = nextElement
          ? this.document.getElementById(nextElement)?.offsetTop
          : fullHeightWithScroll;

        if (fullHeight === scroll && i === offsetHeaders.length) {
          return true;
        }

        return scroll >= Number(offset) && scroll < Number(nextOffset);
      });
    }
  }
```